### PR TITLE
Matching the style with VS MVC Template

### PR DESCRIPTION
**Pull request by omidkrad**
Set width on the form input elements since they're 100% wide by default

### CODE REVIEW

**Comment by omidkrad** [comment]
Do you have any plans for migrating to Bootstrap Sass since looks like everybody is doing that?

**Comment by RehanSaeed** [comment]
Hi Omid, thanks for your comments, I'm about to role them into the next release I will make sometime tomorrow.

I think it is better to use width 100% in conjunction with the Bootstrap [grid system](http://getbootstrap.com/css/#grid), to give a more mobile friendly experience.

I haven't really done any Sass but I promise to look into it and do some reading.

**Comment by omidkrad** [comment]
OK, that's a fair point about the 100% width.

About Sass, checkout this [Google Trend](https://goo.gl/kHldFu) that compares Sass with some other preprocessor Web languages (borrowed from [this](http://developer.telerik.com/featured/the-rise-of-typescript/) article). Also see [this](http://www.zingdesign.com/less-vs-sass-its-time-to-switch-to-sass/) and [this](http://thesassway.com/editorial/sass-vs-scss-which-syntax-is-better). Thanks!

**Comment by RehanSaeed** [comment]
Those are interesting links. I certainly want to learn TS and integrate that into this project but I hadn't considered SASS. I'm not sure that SASS is so popular among Visual Studio developers, I can't find any SASS NuGet packages for Font Awesome.

**Comment by RehanSaeed** [comment]
@omidkrad Bootstrap 4 is going to be SCSS only and ASP.NET 5 has proper support for it using Gulp and Bower. The current unreleased ASP.NET 5 project template is using LESS but I will be transitioning it to SCSS soon. I'll be leaving the ASP.NET 4.6 project template using LESS as there are issues with finding NuGet packages using SCSS.

Thanks for your suggestion of using SCSS.

**Comment by omidkrad** [comment]
You are welcome. Good to know Bootstrap 4 is going to use SCSS only. Sticking to one good solution saves a lot of energy from everyone.

**Comment by omidkrad** [comment]
Hey, I was just looking at new stuff in your [MVC 6 boilerplate](http://rehansaeed.com/whats-new-in-asp-net-5-mvc-6-boilerplate/) and I'm very excited about the new changes. For you to keep your feet on the bleeding edge, I've got a couple suggestions for you. I'd recommend using [jspm](http://jspm.io/) for package management because of its ES6 module support. Also have an eye on the [aurelia](http://aurelia.io) project which is a new client-side framework that I see a great future for it. You may like it so much that you'd want to integrate it in your boilerplate. :smiley:

**Comment by RehanSaeed** [comment]
Thanks, I'll have a read. What are the advantages of these technologies over their competitors. JSPM looks like Bower.

My plan was to add CommonJS module support using Browsify and also support ES6 using TypeScript. Both would be options though in the feature selection wizard.

**Comment by omidkrad** [comment]
It seems JSPM is more standards-oriented than other JS package managers. In my new projects I use ES6 modules only, TypeScript that compiles to ES6 followed by Babel compiling TS outputs to ES5 to run in current browsers.

**Comment by RehanSaeed** [comment]
You can use gulp-typescript to compile TypeScript to ES5 and CommonJS modules too. You don't need Babel I think.

**Comment by RehanSaeed** [comment]
Hmmm, this is an interesting approach. I was thinking about using CommonJS modules with Browserify but ES6 modules with Babel seems a more future proof method. I'm not sure about performance and size of the compiled output though.

I definitely don't want to use AMD/RequireJS because I will have to add another script and we don't really need asynchronous module loading since we are concatenating all the files into one file.

I also don't want to use SystemJS as I don't think we need that much flexibility and I suspect making things that flexible will have some trade off's somewhere. We will also need to add another script.

I want to keep TypeScript as an optional layer on top of this, I'm working on a check-box feature so you can turn it off if you want to.

So in conclusion, I have no idea which option to choose. Mainly because I'm not sure how I feel about forcing users of Boilerplate to learn ES6 if they decide to turn TypeScript off.

As for JSPM vs Bower. I've been doing some reading this morning and I'm not sure about it. It uses SystemJS, which means adding another script.

**Comment by RehanSaeed** [comment]
Ok, I've been thinking about this over lunch. I think the best way to go is give users two options:

1. ES6 JavaScript and ES6 Modules -> Babel -> Single ES5 .js File
2. TypeScript -> Single ES5 .js File

If you are using TypeScript, I'm not sure why you would put Babel into the mix. Am I missing something? Is there a specific reason you added Babel to your Gulp pipeline?

**Comment by omidkrad** [comment]
Exactly. All those module loaders are only transitional solutions, and will go away once there is good support for ES6 on most browsers.

About why I use Babel, you could get TS to compile directly to ES5, however, I had a few roadblocks with that configuration.

- First, when trying to use ES6 module syntax to load some CommonJS modules, even though it compiled, it gave me runtime errors. You can read more about it in this [article](http://www.jbrantly.com/es6-modules-with-typescript-and-webpack/) (see the part about **Importing Libraries**).
- Second, for some of the experimental (ES7) features in typescript, such as `asyncFunctions` TypeScript compiler can only target ES6. And coming from C# world, I cannot miss async/await.

So the current approach, compiling with TS and then with Babel has worked best for me so far. I learned most of these things since I started using Aurelia.

**Comment by RehanSaeed** [comment]
Thanks for the info. async/await is pretty new, I'm sure it will come to ES5 soon enough. If that's the only reason to use Babel with TypeScript then I think it's worth making people wait for TypeScript 1.7 (I don't want too many dependencies that I have to constantly update).

**Comment by omidkrad** [comment]
That makes sense. BTW I have to correct myself here: AMD, CommonJS, UMD are all older and non-standard ways to define/load modules. I think SystemJS is the transitional approach that supports loading those older types of modules and defines new modules in a way that is most compatible with ES6 specs. So I think if you use SystemJS then you won't need other module loaders.

**Comment by omidkrad** [comment]
async/await is a ES7 feature, if it finalizes in the EcmaScript standard. There is a good chance that it will.

**Comment by omidkrad** [comment]
Is there any way to set debug/release modes for JS/TS builds? For example you may want to keep comments and skip uglify/concat in debug mode (or maybe not if using SourceMaps?). I agree in the boilerplate you probably don't want two step compilation with typescript/babel, but surprisingly it works fine with SourceMaps. that is event though the source has gone through TS and babel transpilers, when I debug it in F12 tools , I can step through the original ts code. It blew my mind when I saw that. I currently don't do concat/uglify.

**Comment by omidkrad** [comment]
Sorry since there's no place for discussions I have to write my questions on this unrelated issue ([gitter](http://gitter.im)?)

What does [Boilerplate.Web.Mvc6](https://www.nuget.org/packages/Boilerplate.Web.Mvc6/) package do? Is it a collection of helper files for the boilerplate?

**Comment by RehanSaeed** [comment]
No worries, others can read this conversation. Might be worth raising a new issue for each question.

The NuGet package is basically where I put anything needed that's not part of ASP.NET MVC but should be. Twitter and Facebook HTML/Tag helpers, Sitemap generating code, Filters to redirect to the canonical URL, BSON formatters, AtomActionResult etc.
